plugins/cache-ehcache/src/main/java/org/opensearch/cache/store/disk/EhcacheDiskCache.java

```diff
@@ -135,34 +148,41 @@ private EhcacheDiskCache(Builder<K, V> builder) {
             this.threadPoolAlias = builder.threadPoolAlias;
         }
         this.settings = Objects.requireNonNull(builder.getSettings(), "Settings objects shouldn't be null");
+        this.keySerializer = Objects.requireNonNull(builder.keySerializer, "Key serializer shouldn't be null");
+        this.valueSerializer = Objects.requireNonNull(builder.valueSerializer, "Value serializer shouldn't be null");
         this.cacheManager = buildCacheManager();
-        Objects.requireNonNull(builder.getRemovalListener(), "Removal listener can't be null");
-        this.removalListener = builder.getRemovalListener();
-        this.ehCacheEventListener = new EhCacheEventListener<K, V>(builder.getRemovalListener());
+        this.ehCacheEventListener = new EhCacheEventListener<K, V>(
+            Objects.requireNonNull(builder.getRemovalListener(), "Removal listener can't be null"),
+            Objects.requireNonNull(builder.getWeigher(), "Weigher function can't be null"),
+            this.valueSerializer
+        );
         this.cache = buildCache(Duration.ofMillis(expireAfterAccess.getMillis()), builder);
+        List<String> dimensionNames = Objects.requireNonNull(builder.dimensionNames, "Dimension names can't be null");
+        this.statsHolder = new StatsHolder(dimensionNames, builder.getSettings(), StatsHolder.TrackingMode.ALL_COMBINATIONS);
     }
 
-    private Cache<K, V> buildCache(Duration expireAfterAccess, Builder<K, V> builder) {
+    @SuppressWarnings({ "rawtypes" })
+    private Cache<ICacheKey, byte[]> buildCache(Duration expireAfterAccess, Builder<K, V> builder) {
```

Comment:
I think that dropping the types `K` and `V` is probably a bad idea. 

Isn't this supposed to be a generic cache for any key/value types?

Reply:
There's a couple different (frustrating) limitations in ehcache that make us do this. 

Ideally, we would pass a key and value serializer to ehcache itself, and we wouldn't have to handle serialization ourselves at all. 

But, ehcache requires the exact type of a key/value be maintained after serialization, which wasn't feasible (or necessary) for some interfaces like BytesReference. So, we move the value serializer outside ehcache, and serialize V to/from byte[] before it hits ehcache. 

Ideally we would do this for the key too, but ehcache seems to have a bug when using byte[] as a key. So instead we pass the key serializer to ehcache and accept the limitation that all keys must be the same exact class. 

Finally, now that we have ICache<K> as our key rather than K, we have no way to access the parameterized class object "ICacheKey<String>.class" (or whatever K is in the IRC). We would need this to create an ehcache Cache<ICacheKey<K>, byte[]>. So we have to just do ICacheKey instead. (I could be wrong about this last part -- please let me know if I am!)